cpp/tests/utilities/identify_stream_usage/identify_stream_usage.cu

```diff
@@ -0,0 +1,135 @@
+        // in [begin_offset, end_offset). now apply __cxa_demangle():
+
+        int status;
+        char* ret = abi::__cxa_demangle(begin_name, funcname, &funcnamesize, &status);
+        if (status == 0) {
+          funcname = ret;  // use possibly realloc()-ed string (__cxa_demangle may realloc funcname)
+          std::cout << "#" << i << " in " << strings[i] << " : " << funcname << "+" << begin_offset
+                    << std::endl;
+        } else {
+          // demangling failed. Output function name as a C function with no arguments.
+          std::cout << "#" << i << " in " << strings[i] << " : " << begin_name << "()+"
+                    << begin_offset << std::endl;
+        }
+      } else {
+        std::cout << "#" << i << " in " << strings[i] << std::endl;
+      }
+    }
+
+    free(funcname);
+  }
+  free(strings);
+#else
+  std::cout << "Backtraces are only support on GNU systems." << std::endl;
+#endif  // __GNUC__
+}
+
+// clang-format off
+/*
+   We need to overload all the functions from the runtime API (assuming that we
+   don't use the driver API) that accept streams. Here's a complete listing of
+   the API pages that contain any APIs using streams as of 9/20/2022:
+   - https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__STREAM.html
+   - https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__EVENT.html#group__CUDART__EVENT
+   - https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__EXTRES__INTEROP.html#group__CUDART__EXTRES__INTEROP
+   - https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__EXECUTION.html#group__CUDART__EXECUTION
+   - https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__MEMORY.html#group__CUDART__MEMORY
+   - https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__MEMORY__POOLS.html#group__CUDART__MEMORY__POOLS
+   - https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__OPENGL__DEPRECATED.html#group__CUDART__OPENGL__DEPRECATED
+   - https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__EGL.html#group__CUDART__EGL
+   - https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__INTEROP.html#group__CUDART__INTEROP
+   - https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__GRAPH.html#group__CUDART__GRAPH
+   - https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__HIGHLEVEL.html#group__CUDART__HIGHLEVEL
+ */
+// clang-format on
+
+using cudaLaunchKernel_t = cudaError_t (*)(const void*, dim3, dim3, void**, size_t, cudaStream_t);
+
+static cudaLaunchKernel_t cudaLaunchKernel_original;
+
+void __attribute__((constructor)) init();
+void init()
+{
+  cudaLaunchKernel_original = (cudaLaunchKernel_t)dlsym(RTLD_NEXT, "cudaLaunchKernel");
+}
+
+__host__ cudaError_t cudaLaunchKernel(
+  const void* func, dim3 gridDim, dim3 blockDim, void** args, size_t sharedMem, cudaStream_t stream)
+{
+  if (stream == static_cast<cudaStream_t>(0) || (stream == cudaStreamLegacy) ||
+      (stream == cudaStreamPerThread)) {
```

Comment:
Could this just check if `stream != cudf::get_default_stream().value()` ?

Reply:
It could, and I originally considered that. However, there are two issues:
- That won't necessarily capture usage of `thrust::device` because by default `cudf::default_stream_value` is 0 while (based on my observations) `thrust::device` is `cudaStreamLegacy`(==1).
- That complicates the build because now we need to link this library against libcudf. It's possible, but it feels rather circular to link this against libcudf only to preload it before the run of libcudf. Also, within the context of a CI run where we install conda packages at some point I figured it would make maintenance easier to avoid `find_package(cudf)`.

Reply:
I figured it was something like this. Regardless it may be worth documenting this in the source file for future generations.

Reply:
Done.